src/os/unix/wayland.rs

```diff
@@ -0,0 +1,1182 @@
+                frame[..].as_ptr() as *const u8,
+                frame.len() * std::mem::size_of::<u32>(),
+            )
+        };
+        tmp_f.write_all(&slice[..]).unwrap();
+        tmp_f.flush().unwrap();
+
+        let xdg_wm_base = global_man.instantiate_exact::<XdgWmBase>(1).map_err(|e| {
+            Error::WindowCreate(format!("Failed retrieving the XdgWmBase: {:?}", e))
+        })?;
+
+        //Ping Pong
+        xdg_wm_base.quick_assign(|xdg_wm_base, event, _| {
+            use wayland_protocols::xdg_shell::client::xdg_wm_base::Event;
+
+            if let Event::Ping { serial } = event {
+                xdg_wm_base.pong(serial);
+            }
+        });
+
+        let xdg_surface = xdg_wm_base.get_xdg_surface(&surface);
+        let _surface = surface.clone();
+        //Ping Pong
+        xdg_surface.quick_assign(move |xdg_surface, event, _| {
+            use wayland_protocols::xdg_shell::client::xdg_surface::Event;
+
+            if let Event::Configure { serial } = event {
+                xdg_surface.ack_configure(serial);
+                _surface.commit();
+            }
+        });
+
+        //Assigns the toplevel role and commit
+        let xdg_toplevel = xdg_surface.get_toplevel();
+        if decoration {
+            use wayland_protocols::unstable::xdg_decoration::v1::client::zxdg_decoration_manager_v1::ZxdgDecorationManagerV1;
+
+            if let Ok(deco_man) = global_man
+                .instantiate_exact::<ZxdgDecorationManagerV1>(1)
+                .map_err(|e| {
+                    Error::WindowCreate(format!(
+                        "Failed creating server-side surface decoration: {:?}",
+                        e
+                    ))
+                })
+                .map_err(|e| println!("{:?}", e))
+            {
+                deco_man.get_toplevel_decoration(&xdg_toplevel);
+                deco_man.destroy();
+            }
+        }
+        surface.commit();
+
+        event_q
+            .sync_roundtrip(&mut (), |_, _, _| {})
+            .map_err(|e| Error::WindowCreate(format!("Roundtrip failed: {:?}", e)))?;
+
+        //give the surface the buffer and commit
+        surface.attach(Some(&buffer), 0, 0);
+        surface.damage(0, 0, i32::max_value(), i32::max_value());
+        surface.commit();
+
+        let xdg_config = Rc::new(RefCell::new(None));
+        {
+            let xdg_config = xdg_config.clone();
+            xdg_surface.quick_assign(move |_xdg_surface, event, _| {
+                use wayland_protocols::xdg_shell::client::xdg_surface::Event;
+
+                //Acknowledge only the last configure
+                if let Event::Configure { serial } = event {
+                    *xdg_config.borrow_mut() = Some(serial);
+                }
+            });
+        }
+
+        let cursor = wayland_cursor::load_theme(None, 16, &shm);
+        let cursor_surface = comp.create_surface();
+
+        Ok((
+            Self {
```

Comment:
Since it looks like all these variables are used, there's no need to prefix them with an underscore.